### ISSUE TITLE
maemo-launcher: runs into g++ bug

### ISSUE THREAD

**Issue by MerlijnWajer** [maemo-leste/bugtracker#298]
Compiling maemo-launcher for beowulf (uses `g++-8.3`) gives this error:

```
g++ -DHAVE_CONFIG_H -I. -I..  -fPIC    -c -o booster_cpp_so-booster_cpp.o `test -f 'booster_cpp.cpp' || echo './'`booster_cpp.cpp
In file included from /usr/include/c++/8/map:60,
                 from booster_cpp.cpp:26:
/usr/include/c++/8/bits/stl_tree.h: In instantiation of ‘class std::_Rb_tree<int, std::pair<const int, int>, std::_Select1st<std::pair<const int, int> >, int, std::allocator<std::pair<const int, int> > >’:
/usr/include/c++/8/bits/stl_multimap.h:149:17:   required from ‘class std::multimap<int, int, int>’
booster_cpp.cpp:44:30:   required from here
/usr/include/c++/8/bits/stl_tree.h:452:21: error: static assertion failed: comparison object must be invocable with two arguments of key type
       static_assert(__is_invocable<_Compare&, const _Key&, const _Key&>{},
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gcc -DHAVE_CONFIG_H -I. -I..  -DPROG_NAME="\"defender\"" -DLAUNCHER="\"/usr/bin/maemo-launcher\"" -DDEFENDER="\"/usr/lib/x86_64-linux-gnu/maemo-launcher/defender\""   -Wall -Werror -c -o defender-defender.o `test -f 'defender.c' || echo './'`defender.c
/usr/include/c++/8/bits/stl_tree.h: In instantiation of ‘class std::_Rb_tree<int, int, std::_Identity<int>, int, std::allocator<int> >’:
/usr/include/c++/8/bits/stl_set.h:133:17:   required from ‘class std::set<int, int>’
booster_cpp.cpp:45:30:   required from here
/usr/include/c++/8/bits/stl_tree.h:452:21: error: static assertion failed: comparison object must be invocable with two arguments of key type
gcc -DHAVE_CONFIG_H -I. -I..  -DPROG_NAME="\"defender\"" -DLAUNCHER="\"/usr/bin/maemo-launcher\"" -DDEFENDER="\"/usr/lib/x86_64-linux-gnu/maemo-launcher/defender\""   -Wall -Werror -c -o defender-report.o `test -f 'report.c' || echo './'`report.c
gcc  -Wall -Werror -rdynamic  -o maemo-launcher maemo_launcher-launcher.o maemo_launcher-report.o maemo_launcher-booster.o maemo_launcher-invokelib.o maemo_launcher-comm_msg.o maemo_launcher-comm_dbus.o maemo_launcher-prog.o -ldl -ldbus-1 -ljpeg
make[4]: *** [Makefile:1385: booster_cpp_so-booster_cpp.o] Error 1
make[4]: *** Waiting for unfinished jobs....
```

I can't make sense of it, and neither can this reporter, which resulted in a bug being fixed in `gcc-9.1`: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85965

Compiling maemo-launcher with `g++-7` works fine. Would we be supposed opposed for a patch like this in `maemo/beowulf` branch?

```
diff --git a/debian/control b/debian/control
index d449bde..4d723a8 100644
--- a/debian/control
+++ b/debian/control
@@ -22,6 +22,7 @@ Build-Depends:
  libjpeg62-turbo-dev,
  libsqlite3-dev,
  libalarm-dev,
+ g++-7,
 # To build the QT Booster we need libqt4-dev
 Standards-Version: 4.3.0

diff --git a/debian/rules b/debian/rules
index 7c13546..3ba4d63 100755
--- a/debian/rules
+++ b/debian/rules
@@ -22,7 +22,7 @@ endif
 export DEB_CPPFLAGS_MAINT_APPEND DEB_CFLAGS_MAINT_APPEND

 override_dh_auto_configure:
-       dh_auto_configure -- \
+       CXX="g++-7" dh_auto_configure -- \
                $(opt_hildon) \
                $(opt_qt)
```

cc @spinal84 @freemangordon

**Comment by MerlijnWajer** [comment]
It would result in compiling most of maemo-launcher with gcc-8, but a part with g++-7.

**Comment by spinal84** [comment]
I'd add a comment in both places (debian/control, debian/rules). Something like this:
# g++-8.3 regression workaround (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85965)

**Comment by spinal84** [comment]
Though, I don't think that compiling different parts of package with different compilers is a good idea. It's better to stick on g++-7 for all package.

**Comment by spinal84** [comment]
Anyway, considering the purpose of maemo-launcher, this change should be tested really hard.

**Comment by MerlijnWajer** [comment]
Yes, and until we get to the point where hildon-desktop and osso-xterm is in beowulf, there is nothing we can do to test it. ;)

**Comment by freemangordon** [comment]
> I don't think that compiling parts of a package with different compilers is a good idea. It's better to stick on g++-7 for all the package.

I agree on that one, lets use g++-7 for all of the package, plus, make a note why we need it

**Comment by MerlijnWajer** [comment]
Will do. We'll have to make sure to use this only in beowulf and not ascii, though.

**Comment by MerlijnWajer** [comment]
Closed with `67ede235b2ac45fc7e841fe20119b04ccc81280a`

**Comment by MerlijnWajer** [comment]
Just for future reference, this was definitely fixed in GCC 10, in Devuan Chimaera (Debian Bullseye), so I'm undo'ing the forcing of GCC 7 now.
